File: core/3Sum/3Sum.py
from typing import List

def threeSum(nums: List[int]) -> List[List[int]]:
    nums.sort()
    length = len(nums)
    res = []
    for i in range(length - 2):
        if i != 0 and nums[i] == nums[i - 1]:
            continue
        target = -nums[i]
        third = length - 1
        for j in range(i + 1, length - 1):
            if j != i + 1 and nums[j] == nums[j - 1]:
                continue
            while j < third and nums[j] + nums[third] > target:
                third -= 1
            if j == third:
                break
            elif nums[j] + nums[third] == target:
                res.append([nums[i], nums[j], nums[third]])

    return res


if __name__ == '__main__':
    print(threeSum([-1,0,1,2,-1,-4]))

File: core/3Sum/3Sum.java
package com.dbc;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Sum3 {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i] == nums[i - 1]) continue;
            int target = -nums[i];
            int third = nums.length - 1;
            for (int j = i + 1; j < nums.length - 1; j++) {
                if (j != i + 1 && nums[j] == nums[j - 1]) continue;
                while (j < third && nums[j] + nums[third] > target) third--;
                if (j == third) break;
                else if (nums[j] + nums[third] == target) {
                    List<Integer> temp = new ArrayList<>();
                    temp.add(nums[i]);
                    temp.add(nums[j]);
                    temp.add(nums[third]);
                    res.add(temp);
                }
            }
        }
        return res;
    }
}
